tests/grad_norm_test.py

```diff
@@ -0,0 +1,254 @@
+    len_dataloader = 3
+    max_epochs = max_iterations // len_dataloader + 1
+    norm_vector = []
+
+    for epoch in range(max_epochs):
+        for idx, batch in enumerate(data_loader):
+            step_count = epoch * len_dataloader + idx
+            optimizer.zero_grad()
+
+            logits = model(batch['input_ids'].to('cuda')).logits
+
+            # compute loss
+            loss = _lm_loss(logits, batch['input_ids'].to('cuda').clone())
+            loss.backward()
+
+            if clip_grad:
+                norm_vector.append(nn.utils.clip_grad_norm_(model.parameters(), max_norm))
+            else:
+                norm_vector.append(optimizer.clip_grads_and_collect_metrics(step_count)["total_grad_norm"])
+
+            # apply scheduler
+            for group in optimizer.param_groups:
+                group["lr"] = scheduler.get_lr(cfg.optimizer.learning_rate, step_count, cfg.scheduler.t_max)
+                group["max_grad_norm"] = scheduler.get_max_grad_norm(cfg.max_grad_norm, step_count, cfg.scheduler.t_max)
+                group["max_grad_norm_ratio"] = scheduler.get_max_grad_norm(cfg.max_grad_norm_ratio, step_count, cfg.scheduler.t_max)
+
+            optimizer.step()
+
+            if step_count % 100 == 0:
+                print('Step: {:4d}, Loss: {:.4f}'.format(step_count, loss))
+
+            if step_count == max_iterations:
+                break
+
+    return norm_vector
+
+
+def _run_olmo_grad_norm_againt_torch_grad_norm(
+        rank: int,
+        world_size: int,
+        max_iterations: int,
+        max_norm: float,
+    ):
+    # step 2: run 5 batches on GPU using no_shard fsdp strategy with olmo trainer and compare clipping
+    # step 3: run this for other fsdp strategies
+
+    # minimal distributed env setup
+    # Set up world pg
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    os.environ["RANK"] = str(rank)
+    os.environ["LOCAL_RANK"] = str(rank)
+
+    # Initialize the process group
+    dist.init_process_group("cpu:gloo,cuda:nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+    cfg = TrainConfig.load("test_fixtures/train_tiny.yaml")
+    cfg = _patch_config(cfg, max_norm)
+
+    print('pytorch optim, pytorch grad_clipping, no model wrapping...')
+
+    # run on CPU
+    model = OLMo(cfg.model).to('cuda')
+    torch_optimizer = _init_torch_optim(cfg, model)
+    scheduler = build_scheduler(cfg)
+    data_loader = build_train_dataloader(cfg)
+
+    torch_optim_norms = _naive_train_loop(cfg, model, torch_optimizer, scheduler, data_loader, max_iterations, max_norm, clip_grad=True)
+
+    del model
+    del torch_optimizer
+    del scheduler
+    del data_loader
+
+    # use same model, data, optimizer, fsdp_model and send to trainer and compare gradient clip
+
+    # olmo optimizer
+    model = OLMo(cfg.model).to('cuda')
```

Comment:
Can you make double sure the initialization is same here?

Reply:
From the looks of it, this model will definitely start with different weights than the one above. The best way to be sure would be to load starting state dict of the first model into the second model.